### ISSUE TITLE
Not all files of the MCPileup are added to the configuration

### ISSUE THREAD

**Issue by vlimant** [dmwm/WMCore#7894]
following up on https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4066.html where the software needs all file of the secondary input to be listed in the mix module, @thongonary and I were looking at a specific job log

/eos/cms/store/logs/prod/2017/05/WMAgent/rverma_PR_newco_RelVal_276950_170428_191716_3771/rverma_PR_newco_RelVal_276950_170428_191716_3771-Task1MergeDQMoutputMergedEndOfRunDQMHarvestLogCollect-cmswn2203-1-logs.tar

/WMTaskSpace/logCollect1/2a1532f6-465e-11e7-a342-a0369f23d01e-488-1-logArchive.tar.gz

and the PSet has 2423 files in mixing module while the full dataset does have 5660 files in total, and is in full at the site showing errors about this workflow.

the suspicion is that that secondary file list was truncated at pset tweak time one way or the other.

It's the first time we (at least I) encouter such type of request requiring exactly all files of the secondary to be present.

**Comment by vlimant** [comment]
FYI @prozober

**Comment by vlimant** [comment]
I just looked a the sandbox and the pileupconf.json file.

T1_ES_PIC_Disk 133
T1_FR_CCIN2P3_MSS 5660
T2_CH_CSCS 157
T1_FR_CCIN2P3_Disk 134
T2_TW_NCHC 5660
T2_UK_SGrid_RALPP 60
T2_FR_GRIF_LLR 126
T2_FR_IPHC 226
T0_CH_CERN_Disk 764
T2_DE_DESY 2189
T2_IT_Legnaro 176
T1_DE_KIT_Disk 5660
T2_UK_London_Brunel 2382
T2_IT_Pisa 5660
T1_UK_RAL_Disk 4475
T1_US_FNAL_Disk 124
T2_FR_GRIF_IRFU 104
T1_IT_CNAF_Disk 248
T1_RU_JINR_Disk 282
T2_UK_London_IC 4972
T2_IT_Bari 136
T1_FR_CCIN2P3_Buffer 5660
T2_FR_CCIN2P3 174
T2_ES_CIEMAT 162
T2_US_MIT 5660
T2_BE_IIHE 78
T2_CH_CERN 5660
T2_PT_NCG_Lisbon 65
T2_US_Vanderbilt 348

is the number of files per site, as seen during the sandbox creation.

I can confirm from unified logs that the workflow was assigned with 100% of the dataset at desy (and kit and mit).

So something went astray in creating the sandbox

**Comment by vlimant** [comment]
@ticoann @amaltaro we will need your insight to this, as it has spread over many workflows. Is there anything in the agent log that can give a hint at what went wrong with the sandbx creation ? if not, can we add some logging.

**Comment by amaltaro** [comment]
I had a look at that sandbox and the pileupconf.json file actually contains 2423 files as present at DESY
```
>>> filesInDESY = 0
>>> for block in data['mc']:
...   if 'T2_DE_DESY' in data['mc'][block]['PhEDExNodeNames']:
...     filesInDESY += len(data['mc'][block]['FileList'])
...  
>>> filesInDESY
2423
```

which confirms what you saw in one of the job logs "and the PSet has 2423 files in mixing module".

I was also playing with the phedex APIs  trying to find subscriptions, dataset/block completion and etc and the closest I got was that there was a dataset subscription on the same day that that workflow got assigned (~12h before):
```
Id: 1014153	Requestor: Matteo Cremonesi	Type: xfer	Status: approved	Node: T2_DE_DESY
	Time-created: 2017-05-31 12:57:28	Time-decided: 2017-05-31 12:57:28
```

Thus, my question goes to the transfer team @sidnarayanan and Sebastian (dunno his GH user). Is there any API which tells when a transfer request got 100% completed? Are you able to track when this dataset:
/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM

got completely transferred to T2_DE_DESY?

About the pileup json file, as you could see there are a few PNNs where the dataset is fully available (5660 files), so I doubt there would be any truncation or issue like that building it.

In addition to that, the ACDC workflow prozober_ACDC0_task_HIN-pPb816Summer16DR-00079__v1_T_170607_094539_6258 has all 5660 PU files in the json file (for DESY).

In summary, my suspicious is that something went wrong on 
"I can confirm from unified logs that the workflow was assigned with 100% of the dataset at desy (and kit and mit)." and the data was not completely there, unless Sid can prove me wrong.

**Comment by sidnarayanan** [comment]
Hi Alan,

The best way I know how to tell when it was complete is to check when each of the replicas made it to the endpoint. This can be done via filereplicas [1] and looking at the `time_create` field of the replicas. For this subscription, it looks like the last file replica was created at 1496271816.8.

[1] https://cmsweb.cern.ch/phedex/datasvc/json/prod/filereplicas?dataset=/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM&node=T2_DE_DESY

**Comment by amaltaro** [comment]
Thanks, Sid! Would you by any chance know whether the phedex timestamps are local unix epoch time, or always UTC? BTW, who publishes that replica info? Does it come from every site phedex agent?

Taking into consideration a possible problem with local x utc epoch timestamps, I listed 493 files that were created after that workflow got assigned (1496275109 -7200), using the filereplicas api

Looking at the blockreplicas api and the `time_update` replica attribute, I listed 23 blocks and 989 files, also after the same timestamp. Where several of those blocks were updated yesterday (?)

These numbers don't really match with the lack of ~3k files during the assignment though. However, I'm pretty convinced this dataset wasn't fully available at DESY during the request assignment/acquisition.

**Comment by sidnarayanan** [comment]
Unfortunately I don't have definite answers to any of those 3 questions :). You might want to ask @nataliaratnikova . Certainly the site phedex agent has to tell the central agent that the replica was created, but then I'm not sure if it's the site or central agent that chooses the timestamp. I would hope we use UTC for everything, or things would get confusing...

Sorry, but I'm having a bit of trouble following this. Do I understand correctly:

1.  You tried to run at DESY at some given time, and WMCore saw ~3k missing files at that site at that time

2. You found 493 files (989 files) that were created after the wf got assigned, when checking via filereplicas (blockreplicas). I think 493 vs 989 is not surprising: I assume when you check via blockreplicas, you count all the files in that block if time_update is after your cutoff timestamp. It's possible some of the files were produced before the cutoff, but the block wasn't finished and closed. There could even be a delay between finishing the file injections and closing the block (not sure how you do this in WMCore).

3. There is still the question of why ~3k files were missing. But then I have to ask how you came up with this 2423 number. Is it from PhEDEx, and if so, which API?

Thanks for the clarification.

**Comment by amaltaro** [comment]
No problem, there are too many details and it's easy to get lost.

In summary, when a workflow gets acquired by the agents (after it's assigned in reqmgr2), the agent builds a json for the PU dataset which contains a mapping of blocks x files x PNNs. This json doesn't change during the lifetime of this workflow.

Thus the suspicion is that agents acquired that workflow when not ALL blocks were completed at DESY. That's why we saw 2423 instead of 5660 files at DESY (the agent only considers complete *blocks*)

**Comment by vlimant** [comment]
Wed May 31 23:58:32 2017 (GMT) (consistent in reqmgr2 and unified)

http://dabercro.web.cern.ch/dabercro/unified/showlog/?search=task_HIN-pPb816Summer16DR-00079&module=assignor

https://github.com/CMSCompOps/WmAgentScripts/blob/master/utils.py#L2697

returned 

/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM
{
  "T1_ES_PIC_Disk": [
    false,
    93.3364270657362
...
  "T2_DE_DESY": [
    true,
    100.0
  ],
  "T2_IT_Legnaro": [
    false,
    3.02387718966739
  ],
  "T1_DE_KIT_Disk": [
    true,
    100.0
  ],
...

I cannot see any flaw in the method, and if there was, we'd be seen much more failures all over the place in terms of workflow assigned to site that do not have the data, ... and we do not.

>>> time.asctime(time.gmtime(1496271816.8))
'Wed May 31 23:03:36 2017'
 
 so it looks like the last file replica at DESY was made ~1 hour before the workflow was assigned.

**Comment by vlimant** [comment]
@amaltaro can you please point us to the spot where the PU location is queried ?

**Comment by vlimant** [comment]
note also that the same code, at Wed May 31 22:59:11 2017 returned

  ],
  "T2_DE_DESY": [
    false,
    87.77380013046422
  ],

**Comment by amaltaro** [comment]
Apologies, I missed one important detail:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py#L167

so we do keep a cache of pileup type+dataset in the agent (24h caching).
Apparently there are ~10 workflows in the system using this PU dset.

This PU cache is very important for large campaigns using the same PU and it has a significant impact on how WorkQueueManager acquires work.

Maybe we should decrease it to 1h and live with possible inefficiencies on work acquisition? Of course, there will always be a chance that there is fresh PU data and it uses a stale cache.

**Comment by vlimant** [comment]
I see. Makes sense. I assume you cannot make the cache update conditional on what site is going to be used in sitewhitelist.
from the current crash, 30min would be more adapted, but more impacted.

we could keep the 24h cache lifetime, and in addition have an systematic incremental call to something like 

https://cmsweb.cern.ch/phedex/datasvc/xml/prod/filereplicas?dataset=/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM&node=T2_DE_DESY&since_time=<cache time stamp>

 N.B. since_time would filter on create_time, to update the cache (i.e. we assume that within 24h the dataset is incrementally getting to sites, and cache is flushed every 24h)

Should we get a faster dedicated api from phedex instead ?

**Comment by nataliaratnikova** [comment]
Hi, 
looking at the last_replica time (1496249085) returned by 
https://cmsweb.cern.ch/phedex/datasvc/perl/prod/blocklatencylog?dataset=/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM&to_node=T2_DE_DESY

the last file from this dataset was transferred to DESY on Wed, 31 May 2017 16:44:45 GMT .

**Comment by nataliaratnikova** [comment]
Actually, the above time is the completion time  for some random block I picked from the returned data. The last block in this dataset arrived a few hours later. 
I added  an example: 
https://github.com/nataliaratnikova/PHEDEX/blob/master/Contrib/phedex_dataset_arrival_time_example.ipynb
on how to get the arrival time for the last completed block in this dataset.

**Comment by thongonary** [comment]
@amaltaro A similar issue happened at Purdue: https://hypernews.cern.ch/HyperNews/CMS/get/relval/7838.html

Given that this is a relatively small PU, could you please consider reducing the cache lifetime to a few minutes?

**Comment by amaltaro** [comment]
@thongonary we've just patched the production agents (except those old agents draining). Cache lifetime should be reduced to 30min now.

**Comment by vlimant** [comment]
There is apparently another issue of this type, much longer after the pu timeout being 30min or 24h it seems https://cmsweb.cern.ch/reqmgr2/fetch?rid=vlimant_ACDC0_task_HIN-pPb816Spring16wmLHEGS-00001__v1_T_170721_103006_6723

**Comment by amaltaro** [comment]
Paola provided some ACDC workflows that were created *after* the patch got applied. I picked this one
mcremone_ACDC0_task_HIN-pPb816Spring16wmLHEGS-00006__v1_T_170724_151638_8337-Sandbox.tar.bz2

running on vocms0126 and took a look at its pileupconf.json file (inside the sandbox tarball):
WMSandbox/HIN-pPb816Summer16DR-00147_1/cmsRun1/pileupconf.json

According to DBS, this datasetL
/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM

has 117 blocks, which indeed are all mapped to CERN:
```
>>> with open('pileupconf.json') as jo:
...   data = json.load(jo)   
>>> blocksAtCERN = 0
>>> for block, info in data['mc'].iteritems():
...   if 'T2_CH_CERN' in info['PhEDExNodeNames']:
...     blocksAtCERN += 1
... 
>>> blocksAtCERN
117
```

Thus, I see no issues from the WMCore side. The PU json configuration is correct and contains all blocks.

I don't know how CMSSW decides which event to pick from the pileup dataset (?), but it's probably trying to read an event that does NOT exist at all in the PU sample.

For the record, this is the wmstats job error that I was looking at:
```
Job Name: 2268d5ba-7077-11e7-b4db-02163e0184a6-357
WMBS job id: 1161634
Workflow: mcremone_ACDC0_task_HIN-pPb816Spring16wmLHEGS-00006__v1_T_170724_151638_8337
Task: /mcremone_ACDC0_task_HIN-pPb816Spring16wmLHEGS-00006__v1_T_170724_151638_8337/HIN-pPb816Summer16DR-00147_0
Status: jobfailed
Input dataset:
Site: T2_CH_CERN
Agent: vocms0126.cern.ch
ACDC URL: https://cmsweb.cern.ch/couchdb/acdcserver
State Transition
Exit code: 8026
Retry count: 0
Errors:
    cmsRun1
        Fatal Exception (Exit Code: 8026)

            An exception of category 'NotFound' occurred while
               [0] Processing run: 1 lumi: 17 event: 1441
               [1] Running path 'digitisation_step'
               [2] Calling event method for module MixingModule/'mix'
            Exception Message:
            RootEmbeddedFileSequence::readOneSpecified(): Secondary Input files do not contain specified event:
            run: 1 lumi: 51215 event: 4878241
```

and this lumi section maps to the following LFN:
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/files?logical_file_name=/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/90001/4CE2F999-0009-E711-B779-0025905C2CBC.root&detail=true

which I have no clue whether it contains that event or not? Does anyone know how to list the events in a file? @yuyiguo  ?

**Comment by amaltaro** [comment]
BTW, it was assigned to both CERN and MIT, but MIT doesn't have the complete dataset. It's missing one block.

**Comment by thongonary** [comment]
I checked the file pointed to by Alan and confirmed that it contained the mentioned event:
```
$ root -l /eos/cms/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/90001/4CE2F999-0009-E711-B779-0025905C2CBC.root

root [5] Events->Scan("EventAuxiliary.event()","EventAuxiliary.event()==4878241")
************************
*    Row   * EventAuxi *
************************
*     3345 *   4878241 *
************************
==> 1 selected entry
```
So this doesn't seem to be a dataset's problem either.

**Comment by amaltaro** [comment]
Ok, I have some interesting findings which I can't really explain ATM...Verbose mode ON

Looking at the PU datasets requested for this workflow (yes, there are two different datasets):
https://cmsweb.cern.ch/reqmgr2/data/request?name=mcremone_ACDC0_task_HIN-pPb816Spring16wmLHEGS-00006__v1_T_170724_151638_8337&mask=MCPileup

side note, from my experience, I have had issues in the past when setting AcqEra to `Integ_Test` or `Integ_TEST`. Maybe Yuyi @yuyiguo can confirm whether all dbs operations are case sensitive and thus not an issue in this case.

That task `HIN-pPb816Summer16DR-00147_0` was using the following PU sample:
/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM

which has 117 blocks:
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/filesummaries?dataset=/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM 

Now, if we look at the cmsRun stdout log, this is the file and lumi + event it was looking for
```
24-Jul-2017 22:20:26 UTC  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/00002/A2A83E49-0F08-E711-84C7-02163E013767.root
24-Jul-2017 22:20:28 UTC  Successfully opened file root://eoscms.cern.ch//eos/cms/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/00002/A2A83E49-0F08-E711-84C7-02163E013767.root
----- Begin Fatal Exception 24-Jul-2017 22:20:28 UTC-----------------------
An exception of category 'NotFound' occurred while
   [0] Processing run: 1 lumi: 17 event: 1441
   [1] Running path 'digitisation_step'
   [2] Calling event method for module MixingModule/'mix'
Exception Message:
RootEmbeddedFileSequence::readOneSpecified(): Secondary Input files do not contain specified event:
run: 1 lumi: 51215 event: 4878241
----- End Fatal Exception -------------------------------------------------
24-Jul-2017 22:20:29 UTC  Closed file root://eoscms.cern.ch//eos/cms/store/himc/pPb816Spring16wmLHEGS/DYtoMuMu_M-10to30_pPb-EmbEPOS_8p16_Powheg/GEN-SIM/pPbEmb_80X_mcRun2_pA_v4-v1/130000/0688E6D5-BE6C-E711-88A5-FA163E183C33.root
24-Jul-2017 22:20:29 UTC  Closed file root://eoscms.cern.ch//eos/cms/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/00002/A2A83E49-0F08-E711-84C7-02163E013767.root
MemoryReport> Peak virtual size 1966.58 Mbytes
```

which corresponds to the correct dataset, however this file does NOT have the lumi section 51215
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/filelumis?logical_file_name=/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/00002/A2A83E49-0F08-E711-84C7-02163E013767.root

this lumi section can be found in this file:
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/files?dataset=/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM&run_num=1&lumi_list=[51215]

thus `/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/50001/B496A982-C007-E711-B0AC-1866DA7F90B8.root`

Last but not least, I unpickled the PSet configuration for the job I mentioned in my previous comment  and ... as you can see, the correct file with lumi 51215 is there:
```
amaltaro@lxplus076:~/workarea/PU/CMSSW_8_0_28/src $ grep -r '/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/50001/B496A982-C007-E711-B0AC-1866DA7F90B8.root' myPSet.py 
            '/store/himc/pPb816Spring16GS/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/GEN-SIM/MB_80X_mcRun2_pA_v4-v2/50001/B496A982-C007-E711-B0AC-1866DA7F90B8.root', 
```

Just to be safe, I picked the original (first) workflow sandbox, untarred and checked the pileupconf.json files and indeed the Task1 has a different json compared to Task2 and 3 (which use the same PU sample):
```
cmst1@vocms0116:/data/srv/wmagent/current/alandir $ diff -q ./WMSandbox/HIN-pPb816Summer16DR-00147_0/cmsRun1/pileupconf.json ./WMSandbox/HIN-pPb816Summer16DR-00147_1/cmsRun1/pileupconf.json
cmst1@vocms0116:/data/srv/wmagent/current/alandir $ diff -q ./WMSandbox/HIN-pPb816Spring16wmLHEGS-00006_0/cmsRun1/pileupconf.json ./WMSandbox/HIN-pPb816Summer16DR-00147_1/cmsRun1/pileupconf.json
Files ./WMSandbox/HIN-pPb816Spring16wmLHEGS-00006_0/cmsRun1/pileupconf.json and ./WMSandbox/HIN-pPb816Summer16DR-00147_1/cmsRun1/pileupconf.json differ
```

In summary, I can see 2 problems:
* McM somehow mixed up different PU samples (unless that was intentional)
* the framework got lost on which lumi section should be in a file

**Comment by vlimant** [comment]
indeed the workflow is mis-configured

/ReggeGribovPartonMC_EposLHC_pPb_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM

/ReggeGribovPartonMC_EposLHC_PbP_4080_4080_DataBS/pPb816Spring16GS-MB_80X_mcRun2_pA_v4-v2/GEN-SIM

are requested on the two different steps.
thanks @amaltaro for the paranoiac check. apologies for doubting the "fix"
